common/grpc_client_pool.go

```diff
@@ -0,0 +1,93 @@
+package common
+
+import (
+	"fmt"
+	"sync/atomic"
+
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"google.golang.org/grpc"
+)
+
+// TODO unit test this
+
+// A function that builds a gRPC client of type T.
+type GRPCClientBuilder[T any] func(grpc.ClientConnInterface) T
+
+// GRPCClientPool manages a pool of one or more gRPC clients.
+type GRPCClientPool[T any] struct {
+	// clients is a slice of gRPC clients of type T.
+	clients []T
+
+	// connections is a slice of gRPC client connections. We need to track this in order to be able to close the
+	// connections when the pool is no longer needed.
+	connections []*grpc.ClientConn
+
+	// Incremented once per call to GetClient().
+	callCount atomic.Uint64
+}
+
+// Creates a new GRPCClientPool with the specified client builder and size.
+func NewGRPClientPool[T any](
+	logger logging.Logger,
+	clientBuilder GRPCClientBuilder[T],
+	poolSize int,
+	url string,
+	dialOptions ...grpc.DialOption,
+) (*GRPCClientPool[T], error) {
+
+	if poolSize <= 0 {
+		poolSize = 1
+	}
+
+	// Create the clients up front.
+	connections := make([]*grpc.ClientConn, 0, poolSize)
+	clients := make([]T, 0, poolSize)
+	for i := 0; i < poolSize; i++ {
+		conn, err := grpc.NewClient(url, dialOptions...)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create gRPC client connection to %s: %w", url, err)
```

Comment:
Note: I really hate that the golang library decided to use "connection" instead of "channel", because I was confused up until today and thought that this NewClient was actually creating an underlying physical tcp connection. But it's actually not! Those are lazily created when sending the first grpc....

I think all that this does here is do name resolution, so this could fail if the dns name doesn't resolve to an ip for example.